# Lab 5: Project Roadmap

In this assignment, you will advance your course project by creating a roadmap of tasks and performing a critical exploratory analysis of your dataset.

In [57]:
# --- Import Libraries ---
# additional libraries need to be added to the requirements.txt file (see README.md for more details)
import pandas as pd
import numpy as np

## Part 1: Data Ingestion & Reality Check (1.5 Points)

Your first task is to load the data and verify that it matches your mental model of the dataset.

In [58]:
df = pd.read_csv("../data/AB_NYC_2019.csv")


In [59]:
df.shape


(48895, 16)

In [60]:
df.columns = (
    df.columns
      .str.lower()
      .str.strip()
      .str.replace(" ", "_")
)

# Example: df = df.rename(columns={'Old Name': 'new_name'})

In [61]:
df.head()
df.tail()


# Example: df['date_col'] = pd.to_datetime(df['date_col'])

,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
48890,36484665,Charming one bedroom - newly renovated rowhouse,8232441,Sabrina,Brooklyn,Bedford-Stuyvesant,40.67853,-73.94995,Private room,70,2,0,NaN,NaN,2,9
48891,36485057,Affordable room in Bushwick/East Williamsburg,6570630,Marisol,Brooklyn,Bushwick,40.70184,-73.93317,Private room,40,4,0,NaN,NaN,2,36
48892,36485431,Sunny Studio at Historical Neighborhood,23492952,Ilgar & Aysel,Manhattan,Harlem,40.81475,-73.94867,Entire home/apt,115,10,0,NaN,NaN,1,27
48893,36485609,43rd St. Time Square-cozy single bed,30985759,Taz,Manhattan,Hell's Kitchen,40.75751,-73.99112,Shared room,55,1,0,NaN,NaN,6,2
48894,36487245,Trendy duplex in the very heart of Hell's Kitchen,68119814,Christophe,Manhattan,Hell's Kitchen,40.76404,-73.98933,Private room,90,7,0,NaN,NaN,1,23


### Part 1 Analysis
1. Does the row count match your expectation?
The row count matches expectations. The dataset contains 48,895 rows and 16 columns, which is consistent with a large public dataset of Airbnb listings in New York City. This sample size is sufficient for meaningful exploratory analysis.
2. Did the column renaming work?
The column renaming worked successfully. All column names are now lowercase and use underscores instead of spaces, which makes them easier to reference in Python. 
3. Have you changed any of the variable data types?
Yes, the `last_review` column was converted from a string (object) to a datetime data type using `pd.to_datetime()`. This allows for proper time-based analysis in later stages of the project.

In [62]:
df.head(10)
df.tail()


,id,name,host_id,host_name,neighbourhood_group,neighbourhood,latitude,longitude,room_type,price,minimum_nights,number_of_reviews,last_review,reviews_per_month,calculated_host_listings_count,availability_365
48890,36484665,Charming one bedroom - newly renovated rowhouse,8232441,Sabrina,Brooklyn,Bedford-Stuyvesant,40.67853,-73.94995,Private room,70,2,0,NaN,NaN,2,9
48891,36485057,Affordable room in Bushwick/East Williamsburg,6570630,Marisol,Brooklyn,Bushwick,40.70184,-73.93317,Private room,40,4,0,NaN,NaN,2,36
48892,36485431,Sunny Studio at Historical Neighborhood,23492952,Ilgar & Aysel,Manhattan,Harlem,40.81475,-73.94867,Entire home/apt,115,10,0,NaN,NaN,1,27
48893,36485609,43rd St. Time Square-cozy single bed,30985759,Taz,Manhattan,Hell's Kitchen,40.75751,-73.99112,Shared room,55,1,0,NaN,NaN,6,2
48894,36487245,Trendy duplex in the very heart of Hell's Kitchen,68119814,Christophe,Manhattan,Hell's Kitchen,40.76404,-73.98933,Private room,90,7,0,NaN,NaN,1,23


### Visual Inspection

The headers appear correctly formatted after cleaning, with consistent lowercase
naming and underscores. The data at the bottom of the file appears consistent
with the top of the dataset, indicating no truncation or footer artifacts.

## Part 2: Deep Dive & Data Dictionary (2 Points)
You will now analyze the content of each column to build a "Data Dictionary" and assess data quality

In [63]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  str    
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  str    
 4   neighbourhood_group             48895 non-null  str    
 5   neighbourhood                   48895 non-null  str    
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  str    
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     38843 non-n

In [64]:
df.describe()


,id,host_id,latitude,longitude,price,minimum_nights,number_of_reviews,reviews_per_month,calculated_host_listings_count,availability_365
count,4.889500e+04,4.889500e+04,48895.000000,48895.000000,48895.000000,48895.000000,48895.000000,38843.000000,48895.000000,48895.000000
mean,1.901714e+07,6.762001e+07,40.728949,-73.952170,152.720687,7.029962,23.274466,1.373221,7.143982,112.781327
std,1.098311e+07,7.861097e+07,0.054530,0.046157,240.154170,20.510550,44.550582,1.680442,32.952519,131.622289
min,2.539000e+03,2.438000e+03,40.499790,-74.244420,0.000000,1.000000,0.000000,0.010000,1.000000,0.000000
25%,9.471945e+06,7.822033e+06,40.690100,-73.983070,69.000000,1.000000,1.000000,0.190000,1.000000,0.000000
50%,1.967728e+07,3.079382e+07,40.723070,-73.955680,106.000000,3.000000,5.000000,0.720000,1.000000,45.000000
75%,2.915218e+07,1.074344e+08,40.763115,-73.936275,175.000000,5.000000,24.000000,2.020000,2.000000,227.000000
max,3.648724e+07,2.743213e+08,40.913060,-73.712990,10000.000000,1250.000000,629.000000,58.500000,327.000000,365.000000


In [65]:
df.describe(include='object')


/tmp/ipykernel_3685/4084012645.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include='object')


,name,host_name,neighbourhood_group,neighbourhood,room_type,last_review
count,48879,48874,48895,48895,48895,38843
unique,47896,11452,5,221,3,1764
top,Hillside Hotel,Michael,Manhattan,Williamsburg,Entire home/apt,6/23/2019
freq,18,417,21661,3920,25409,1413


In [66]:
df.isna().sum()


id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

## Data Dictionary & Quality Check

### Column: price
- **Represents:** Nightly listing price in USD.
- **Data Type:** Numeric (int64) — correct.
- **Sanity Check:** Minimum value is 0 (suspicious). Maximum value is very high, indicating possible outliers.
- **Missing Data Plan:** No missing values. Outliers will be reviewed later.

---

### Column: neighbourhood_group
- **Represents:** NYC borough (Manhattan, Brooklyn, etc.).
- **Data Type:** Object (string) — correct.
- **Sanity Check:** Values match expected borough names.
- **Missing Data Plan:** No missing values.

---

### Column: room_type
- **Represents:** Type of rental (Entire home/apt, Private room, Shared room).
- **Data Type:** Object (string) — correct.
- **Sanity Check:** Only 3 categories exist, which is reasonable.
- **Missing Data Plan:** No missing values.

---

### Column: last_review
- **Represents:** Date of most recent review.
- **Data Type:** Datetime — correctly converted.
- **Sanity Check:** Dates appear valid.
- **Missing Data Plan:** 10,052 missing values likely mean listings with no reviews. These will not be dropped but handled appropriately later.

---

### Column: reviews_per_month
- **Represents:** Average number of reviews per month.
- **Data Type:** Float — correct.
- **Sanity Check:** Values are non-negative.
- **Missing Data Plan:** 10,052 missing values correspond to listings with no reviews. These may be filled with 0.


## Part 3: Project Strategy (1.5 Points)
Now that you have critically assessed your data quality, you must map it to your research goals that you indicated in Project Assignment #1.

## Research Questions

1. How does listing price vary across different NYC boroughs?
   - Columns needed: price, neighbourhood_group

2. Does room type significantly impact listing price?
   - Columns needed: price, room_type

3. Is there a relationship between availability and listing price?
   - Columns needed: price, availability_365


**Statistical Analysis Plan:** *Double click here to edit*

*Explain which statistical test (t-test, regression, etc.) you will use for each question and WHY.*

In [67]:
project_df = df[['price', 'neighbourhood_group', 'room_type', 'availability_365']]

project_df.head()


,price,neighbourhood_group,room_type,availability_365
0,149,Brooklyn,Private room,365
1,225,Manhattan,Entire home/apt,355
2,150,Manhattan,Private room,365
3,89,Brooklyn,Entire home/apt,194
4,80,Manhattan,Entire home/apt,0
